pkg/util/resize/configspec_devices.go

```diff
@@ -25,16 +26,633 @@ func compareHardwareDevices(
+		dc := zipVirtualDevicesOfType(MatchVirtualPrecisionClock, exp, cur)
+		deviceChanges = append(deviceChanges, dc...)
+	}
+
+	{
+		exp, cur := selectDevsByType[*vimtypes.VirtualSCSIPassthrough](expectedDevices, currentDevices)
+		dc := zipVirtualDevicesOfType(MatchVirtualSCSIPassthrough, exp, cur)
+		deviceChanges = append(deviceChanges, dc...)
+	}
+
+	{
+		exp, cur := selectDevsByType[*vimtypes.VirtualSerialPort](expectedDevices, currentDevices)
+		dc := zipVirtualDevicesOfType(MatchVirtualSerialPort, exp, cur)
+		deviceChanges = append(deviceChanges, dc...)
+	}
+
+	{
+		exp, cur := selectDevsByType[*vimtypes.VirtualEnsoniq1371](expectedDevices, currentDevices)
+		dc := zipVirtualDevicesOfType(MatchVirtualEnsoniq1371, exp, cur)
+		deviceChanges = append(deviceChanges, dc...)
+	}
+
+	{
+		exp, cur := selectDevsByType[*vimtypes.VirtualHdAudioCard](expectedDevices, currentDevices)
+		dc := zipVirtualDevicesOfType(MatchVirtualHdAudioCard, exp, cur)
+		deviceChanges = append(deviceChanges, dc...)
+	}
+
+	{
+		exp, cur := selectDevsByType[*vimtypes.VirtualSoundBlaster16](expectedDevices, currentDevices)
+		dc := zipVirtualDevicesOfType(MatchVirtualSoundBlaster16, exp, cur)
+		deviceChanges = append(deviceChanges, dc...)
+	}
+
+	{
+		exp, cur := selectDevsByType[*vimtypes.VirtualTPM](expectedDevices, currentDevices)
+		dc := zipVirtualDevicesOfType(MatchVirtualTPM, exp, cur)
+		deviceChanges = append(deviceChanges, dc...)
+	}
+
+	{
+		exp, cur := selectDevsByType[*vimtypes.VirtualWDT](expectedDevices, currentDevices)
+		dc := zipVirtualDevicesOfType(MatchVirtualWDT, exp, cur)
+		deviceChanges = append(deviceChanges, dc...)
+	}
+
+	return deviceChanges
+}
+
+func selectDevsByType[T vimtypes.BaseVirtualDevice](
+	expectedDevices, currentDevices []vimtypes.BaseVirtualDevice) ([]T, []T) {
+
+	return pkgutil.SelectDevicesByType[T](expectedDevices), pkgutil.SelectDevicesByType[T](currentDevices)
+}
+
+func zipVirtualDevicesOfType[T vimtypes.BaseVirtualDevice](
+	matchDevFn func(expectedDev, curDev T) vimtypes.BaseVirtualDevice,
+	expectedDevs, curDevs []T) []vimtypes.BaseVirtualDeviceConfigSpec {
+
+	var deviceChanges []vimtypes.BaseVirtualDeviceConfigSpec //nolint:prealloc
+
+	// For the expected and current devices of a type, zip the devices together to determine
+	// what, if any, edits need to be done. Then add and remove what is leftover.
+	minLen := min(len(expectedDevs), len(curDevs))
+	for idx := range minLen {
+		expectedDev, curDev := expectedDevs[idx], curDevs[idx]
+
+		editDev := matchDevFn(expectedDev, curDev)
+		if editDev != nil {
+			deviceChanges = append(deviceChanges, &vimtypes.VirtualDeviceConfigSpec{
+				Operation: vimtypes.VirtualDeviceConfigSpecOperationEdit,
+				Device:    editDev,
+			})
+		}
+	}
+
+	// Add new expected devices.
+	for idx := range expectedDevs[minLen:] {
+		deviceChanges = append(deviceChanges, &vimtypes.VirtualDeviceConfigSpec{
```

Comment:
Prb wanna ignore defaultDevice 'Adds' too?. (or) do we wanna let it go to vpxd and let it drop

Reply:
I'd prefer to let it go. First, if a default device somehow got removed and it is in the class's ConfigSpec we should try to re-add it. Second, and most common would be that the class's ConfigSpec is invalid in that it has multiple adds for a default device. I think we should still add it and let it fail, since things can already fail for a million other reasons.

Reply:
makes sense. 
Just to put out there. 
AFAIK,  In the first case, the re-add will be ignored by vpxd in the 'reconfigure' call. So it won't make a difference and we can let vpxd make the call.  
The second case has potential for failure if invalid, true.